File: Python/09_Spiral Matrix.py
# 54. Spiral Matrix
# https://leetcode.com/problems/spiral-matrix/description/

class Solution:
    def spiralOrder(self, matrix):
        def sprial(l):
            ans=[]
            row,col=len(l),len(l[0])
            for i in range(col):
                ans.append(l[0][i])
            for i in range(1,row):
                ans.append(l[i][col-1])
            if row!=1:
                for i in range(col-2,-1,-1):
                    ans.append(l[row-1][i])
            if col!=1:
                for i in range(row-2,0,-1):
                    ans.append(l[i][0])
            return ans

        ans=sprial(matrix)
        temp=matrix.copy()
        while temp:
            l=[]
            try:
                temp.pop()
                temp.pop(0)
                for i in range(len(temp)):
                    temp[i].pop(0)
                    temp[i].pop()
                l=sprial(temp)
                for i in l:
                    ans.append(i)
            except:
                pass
        return ans

File: C++/09_Spiral_Matrix.cpp
//54. Spiral Matrix
//Problem Link :- https://leetcode.com/problems/spiral-matrix/description/

//Approach :-
/*
    To traverse the matrix, we can do it in 4 steps. But we need to take care of the fact
    that after a cycle of these 4 steps, the matrix left to traverse gets smaller. Hence we 
    make 4 pointers to define the current boundary of matrix.
    cleft = left boundary(intially 0), cright = right boundary(intially m-1),
    rup = upper boundary(intially 0), rdown = lower boudndary(initially n-1)
    The  steps to traverse the matrix are :-
    1. Traverse from matrix[rup][cleft] to matrix[rup][cright] and then increment rup.
    2. Traverse from matrix[rup][cright] to matrix[rdown][cright] and decrement cright.
    3. Traverse from matrix[rdown][cright] to matrix[rdown][cleft] and and increment rdown.
    4. Traverse from matrix[rdown][cleft] to matrix[rup][cright] and increment cleft.

    We perform these 4 steps until we have completely traversed the matrix.

*/

//Code :-
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int n = (int)matrix.size();
        int m = (int)matrix[0].size();
        int cleft = 0,cright = m-1;
        int rup = 0,rdown = n-1;
        vector<int>ans;
        while(cleft<=cright && rup<=rdown){
            for(int i=cleft;i<=cright;i++){
                ans.push_back(matrix[rup][i]);
            }
            rup++;
            if(rup>rdown) break;
            for(int i=rup;i<=rdown;i++){
                ans.push_back(matrix[i][cright]);
            }
            cright--;
            if(cright<cleft) break;
            for(int i=cright;i>=cleft;i--){
                ans.push_back(matrix[rdown][i]);
            }
            rdown--;
            if(rdown<rup) break;
            for(int i=rdown;i>=rup;i--){
                ans.push_back(matrix[i][cleft]);
            }
            cleft++;
        }
        return ans;
    }
};